keps/sig-cli/4764-kubectl-watchlist/README.md

```diff
@@ -0,0 +1,757 @@
+
+In practice, our users have been writing script to unrestrainedly call `kubectl`, which could lead serious performance
+issue. As cluster admininstor, we have few method to eliminate it. APF maybe is an option, but it's hard to configure
+and don't solve the problem from the root. In the other side, if we are standing in the shoes of user, most of those 
+scripts are simple Bash or Python which just simply glue a few `kubectl` commands. It is too heavy and "not worth" 
+for user to refactor just "a few commands" into "a batch of complex code". So we are seeking an optimization for `kubectl`. With this optimization, all users of kubectl can get free performance improvement.
+
+
+
+### Goals
+
+* Introduce WatchList to `kubectl` without breaking any user interface.
+* Make `kubectl` hit the cache, thus reducing impact on `kube-apiserver`.
+
+### Non-Goals
+
+* Change kubectl functionality.
+
+## Proposal
+
+`kubectl` commands like `get`, `describe` now should be able to use `WatchList` to list resources, rather than
+directly list resources page by page.
+
+At this stage, we will adopt a similar approach to `client-go`, using an environment variable to control whether
+to enable this experimental feature, and ensuring that it can revert to the previous behavior if it fails.
+
+<<[UNRESOLVED Will `WatchList` replace `List` completely]>>
+`WatchList` was first released in version 1.27 and is still in the alpha stage, with continuous improvements 
+over the subsequent releases. For a feature in such an early stage, it is too soon to say whether we will fully
+rely on it without other fallback mechanisms.
+
+In my view, due to the performance improvements, we should use `WatchList` by default instead of `List` in the
+long term. For the evolution strategy, we can follow the approach of `client-go` and address it in a later release.
+
+<<[/UNRESOLVED]>>
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a user, I shouldn't need to worry about the underlying details. I should simply update the version of `kubectl`
+to gain the performance improvements for free.
+
+#### Story 2
+
+As a cluster administrator, after upgrading the cluster, I should observe a significant reduction in the requests
+penetrating into etcd on the monitoring dashboard. Ideally, the majority of the requests should be cached by the 
+apiserver, rather than penetrating into etcd.
+
+### Risks and Mitigations
+
+For users, there are no visible changes to the user interface. However, since different APIs are called, 
+some users with restricted permissions may encounter authorization issues. Specifically, users might need
+to be granted additional `watch` permissions for the corresponding resources to ensure this mechanism 
+works properly. Fortunately, we have a fallback mechanism in this place. Even if users do not have the 
+additional permissions, the old behavior should still function properly. In the worst case, some extra API 
+calls might cause a slight delay.
```

Comment:
Will kubectl cache a hint about whether or not a cluster supports **watch** for fetching collections?

Reply:
I think it is possible to do so by cache it into a global variable. In this way, during the life cycle of the kubectl process, if we find that **watchlist** is not supported, we can directly skip using the **watchlist** mechanism later.

Reply:
Ah, I meant a (hint) cache that could outlast a particular kubectl process.

Reply:
Yeah, that's a great idea. maybe we can write a cache file to record if current cluster support **watchlist** or not.  this cache file would be valid for a short period of time afterwards. so other kubectl process could read it and fallback to **list** directly without trying **watchlist**.

But my concern is in the implementation. Currently, there are many references to List (https://github.com/kubernetes/kubernetes/blob/fe3772890f650f9bcf020b43dc5a51fab0fa17f4/staging/src/k8s.io/cli-runtime/pkg/resource/helper.go#L64) and they are deeply nested. Most of them do not involve the acquisition of the cache directory. If we want to add it, we need to change a lot of places. This may cause some omissions and lead to some problems. Maybe we don’t have to keep it exactly the same as the previous cache directory, but find a separate directory?